File: dto/package.go
package dto

import (
	"GoPkgRepository/types"
)

type RepositoryType string

const (
	RepositoryTypeGit        RepositoryType = "git"
	RepositoryTypeMercurial  RepositoryType = "hg"
	RepositoryTypeSubversion RepositoryType = "svn"
	RepositoryTypeFossil     RepositoryType = "fossil"
)

type Package struct {
	ID int `json:"id"`

	ImportPath    string         `json:"import_path"`
	VCS           RepositoryType `json:"vcs"`
	RepositoryURL *types.URI     `json:"repository_url"`

	SourceURL     *types.URI `json:"source_url"`
	SourceDirURL  *types.URI `json:"source_dir_url"`
	SourceFileURL *types.URI `json:"source_file_url"`
}

func (receiver Package) IsValid() bool {
	return receiver.ImportPath != "" && receiver.VCS != "" && receiver.RepositoryURL != nil
}

File: ui/src/app/dto/package.ts
export enum Vcs {
  Git = "git",
  Mercurial = "hg",
  Subversion = "svn",
  Fossil = "fossil",
}

export interface Package<TNullType = null> {
  id: number;

  import_path: string;
  vcs: Vcs;
  repository_url: string;

  source_url: string | TNullType;
  source_dir_url: string | TNullType;
  source_file_url: string | TNullType;
}

export const emptyPackage = <TNullType>(nullValue: TNullType): Package<TNullType> => ({
  id: 0,
  import_path: '',
  vcs: Vcs.Git,
  repository_url: '',
  source_url: nullValue,
  source_dir_url: nullValue,
  source_file_url: nullValue,
});
